src/OpenLoco/src/Graphics/SoftwareDrawingContext.cpp

```diff
@@ -726,16 +726,243 @@ namespace OpenLoco::Gfx
+                {
+                    return;
+                }
+                imageHeight -= v37;
+            }
+
+            int16_t rowSize = 0;
+            int16_t dstWrap = rt.pitch + (rt.width >> TZoomLevel);
+
+            if constexpr (TZoomLevel == 0)
+            {
+                dstWrap -= g1Image->width;
+            }
+
+            int16_t rtPosX = ((g1Image->xOffset + pos.x + offsetX) & zoomMask) - rt.x;
+            if (rtPosX < 0)
+            {
+                if (rtPosX + imageWidth == 0)
+                {
+                    return;
+                }
+
+                imageWidth += rtPosX;
+
+                if (imageWidth < 0)
+                {
+                    return;
+                }
+
+                rowSize -= rtPosX;
+                imageDataPos -= rtPosX;
+                if constexpr (TZoomLevel == 0)
+                {
+                    dstWrap -= rtPosX;
+                }
+
+                rtPosX = 0;
+            }
+
+            int16_t rtPosXUnscaled = rtPosX;
+            rtPosX = rtPosX >> TZoomLevel;
+
+            int16_t v42 = imageWidth + rtPosXUnscaled - rt.width;
+            if (imageWidth + rtPosXUnscaled > rt.width)
+            {
+                if (imageWidth <= v42)
+                {
+                    return;
+                }
+
+                imageWidth -= v42;
+                rowSize += v42;
+
+                if constexpr (TZoomLevel == 0)
+                {
+                    dstWrap += v42;
+                }
+            }
+
+            dstBuf += rtPosX;
+
+            auto imageOffset = imageDataPos - g1Image->offset;
+            drawMaskedZoom<TZoomLevel>(
+                imageHeight,
+                imageWidth,
+                rowSize,
+                &g1ImageMask->offset[imageOffset],
+                dstWrap,
+                dstBuf,
+                imageDataPos);
+        }
+
         // 0x00450705
         static void drawImageMasked(const RenderTarget& rt, const Ui::Point& pos, const ImageId& image, const ImageId& maskImage)
         {
-            registers regs;
-            regs.edi = X86Pointer(&rt);
-            regs.cx = pos.x;
-            regs.dx = pos.y;
-            regs.ebx = image.toUInt32();
-            regs.ebp = maskImage.toUInt32();
-            call(0x00450705, regs);
+            switch (rt.zoomLevel)
+            {
+                case 0:
+                    drawImageMaskedZoom<0>(rt, pos, image, maskImage);
+                    break;
+                case 1:
+                    drawImageMaskedZoom<1>(rt, pos, image, maskImage);
+                    break;
+                case 2:
+                    drawImageMaskedZoom<2>(rt, pos, image, maskImage);
+                    break;
+                case 3:
+                    drawImageMaskedZoom<3>(rt, pos, image, maskImage);
+                    break;
+                default:
+                    break;
+            }
+            // Unreachable
```

Comment:
This is reachable the function doesn't return early perhaps you meant to put this in the default

Reply:
You are right, I meant to have return and not break, ill change it.